clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ModifyServerGroupLaunchTemplateAtomicOperation.java

```diff
@@ -0,0 +1,251 @@
+        description.getAssociatePublicIpAddress() != null
+            && description.getAssociatePublicIpAddress();
+    if (networkInterface != null) {
+      if (associatePublicIp) {
+        // associate ipV6 address
+        networkInterface.withIpv6AddressCount(1);
+      }
+
+      networkInterface.withAssociatePublicIpAddress(associatePublicIp).withGroups(securityGroups);
+    } else {
+      // associate ipV6 address
+      LaunchTemplateInstanceNetworkInterfaceSpecificationRequest newInterface =
+          new LaunchTemplateInstanceNetworkInterfaceSpecificationRequest()
+              .withDeviceIndex(0)
+              .withGroups(securityGroups)
+              .withAssociatePublicIpAddress(associatePublicIp);
+
+      if (associatePublicIp) {
+        newInterface.withIpv6AddressCount(1);
+      }
+
+      newLaunchTemplateData.getNetworkInterfaces().add(newInterface);
+    }
+
+    String requireIMDSv2 =
+        description.getRequireIMDV2() != null && description.getRequireIMDV2() ? "required" : "";
+    newLaunchTemplateData.withMetadataOptions(
+        new LaunchTemplateInstanceMetadataOptionsRequest().withHttpEndpoint(requireIMDSv2));
+
+    // if we are changing instance types and don't have explicitly supplied block device mappings
+    String instanceType = description.getInstanceType();
+    List<AmazonBlockDevice> blockDevices = null;
+    if (description.getBlockDevices() == null
+        && instanceType != null
+        && !instanceType.equals(launchTemplateData.getInstanceType())) {
+      List<AmazonBlockDevice> blockDevicesForInstanceType =
+          blockDeviceConfig.getBlockDevicesForInstanceType(instanceType);
+      if (!description.getCopySourceCustomBlockDeviceMappings()) {
+        blockDevices = blockDevicesForInstanceType;
+      } else {
+        // if prior version used default mapping do use default mapping on new version as well
+        List<LaunchTemplateBlockDeviceMapping> currentBlockDeviceMappings =
+            currentVersion.getLaunchTemplateData().getBlockDeviceMappings();
+
+        for (LaunchTemplateBlockDeviceMapping mapping : currentBlockDeviceMappings) {
+          if (blockDevicesForInstanceType.stream()
+              .anyMatch(deviceForType -> matchesDevice(deviceForType, mapping))) {
+            blockDevices = blockDevicesForInstanceType;
+            break;
+          }
+        }
+      }
+    }
+
+    LaunchTemplateVersion newVersion =
+        launchTemplateService.modifyLaunchTemplate(
+            currentVersion,
+            description.getBase64UserData(),
+            newLaunchTemplateData,
+            Optional.of(blockDevices));
+
+    launchTemplateSpec.setVersion(String.valueOf(newVersion.getVersionNumber()));
+
+    // point asg to new template version
+    autoScaling.updateAutoScalingGroup(
+        new UpdateAutoScalingGroupRequest()
+            .withAutoScalingGroupName(description.getAsgName())
+            .withLaunchTemplate(launchTemplateSpec));
```

Comment:
Are there reasons why we _aren't_ using sagas for atomic operations that make more than one write operation to AWS?

Reply:
Definitely a saga opportunity.